File: clients/r/R/wire.R
#' Unwrap one tagged `Value` cell (contract Sec.3) into a plain R value.
#'
#' @param value A parsed JSON value (as a nested list, via
#'   `httr2::resp_body_json(resp, simplifyVector = FALSE)`).
#' @return `NA`, a scalar, a numeric vector (Vector), or a list of numeric
#'   vectors (Matrix, one per row).
#' @keywords internal
#' @noRd
unwrap_value <- function(value) {
  if (identical(value, "Null")) {
    return(NA)
  }
  if (is.list(value) && length(value) == 1 && !is.null(names(value))) {
    key <- names(value)[[1]]
    inner <- value[[1]]
    if (key == "Float") {
      return(as.numeric(inner))
    }
    if (key == "Int") {
      return(as.integer(inner))
    }
    if (key == "String") {
      return(as.character(inner))
    }
    if (key == "Bool") {
      return(as.logical(inner))
    }
    if (key == "Vector") {
      return(as.numeric(unlist(inner)))
    }
    if (key == "Matrix") {
      return(lapply(inner, function(row) as.numeric(unlist(row))))
    }
  }
  stop(linal_error(paste0(
    "Unrecognized Value wire shape: ",
    jsonlite::toJSON(value, auto_unbox = TRUE)
  )))
}

#' Whether an unwrapped `Value` is scalar (safe to put in an atomic vector
#' column) vs. a Vector/Matrix (must stay a list-column). Note: a
#' single-element Vector column (`Vector(1)`) is indistinguishable here from
#' a Float column and gets flattened too -- the numeric content is
#' unaffected, only the "this was semantically a Vector" tag is lost for
#' that degenerate case. A column mixing 1-element and longer vectors still
#' correctly stays a list-column, since not every row would be scalar then.
#' @keywords internal
#' @noRd
is_scalar_value <- function(v) {
  length(v) == 1 && !is.list(v)
}

#' Convert a real `Table` payload (contract Sec.1's verified shape --
#' `payload$schema$fields` for column order/names, `payload$rows[[i]]$values`
#' for each row's cells, NOT the row object itself) into a named list of
#' per-row-unwrapped-value lists, one per column.
#' @keywords internal
#' @noRd
table_result_columns <- function(payload) {
  fields <- payload$schema$fields
  col_names <- vapply(fields, function(f) f$name, character(1))
  rows <- payload$rows
  ncols <- length(col_names)

  columns <- vector("list", ncols)
  names(columns) <- col_names
  for (j in seq_len(ncols)) {
    columns[[j]] <- lapply(rows, function(r) unwrap_value(r$values[[j]]))
  }
  columns
}

#' Convert `table_result_columns()`'s output into a `data.frame`. A column
#' where every value is scalar becomes a normal atomic vector column; a
#' column containing any Vector/Matrix value (or a mix with `NA`) stays a
#' list-column, R's standard way to hold non-atomic per-row values in a
#' data.frame.
#' @keywords internal
#' @noRd
columns_to_dataframe <- function(columns) {
  if (length(columns) == 0) {
    return(data.frame())
  }
  nrows <- length(columns[[1]])
  df <- data.frame(row.names = seq_len(nrows))
  for (name in names(columns)) {
    col <- columns[[name]]
    if (all(vapply(col, is_scalar_value, logical(1)))) {
      df[[name]] <- unlist(lapply(col, function(v) if (length(v) == 1 && is.na(v)) NA else v))
    } else {
      df[[name]] <- col
    }
  }
  df
}

#' Unwrap one `DslOutput` JSON value (contract Sec.1) into `NULL` (no
#' output), a `character` scalar (`Message`), a `linal_table_result`, or a
#' `linal_tensor_result`.
#' @keywords internal
#' @noRd
unwrap_result <- function(result) {
  if (is.null(result)) {
    return(NULL)
  }
  if (!is.list(result) || length(result) != 1 || is.null(names(result))) {
    stop(linal_error("Unrecognized DslOutput wire shape"))
  }
  kind <- names(result)[[1]]
  payload <- result[[1]]

  if (kind == "Message") {
    return(payload)
  }
  if (kind == "Table") {
    return(structure(
      list(columns = table_result_columns(payload)),
      class = "linal_table_result"
    ))
  }
  if (kind == "TensorTable") {
    # Not yet exercised against a real response -- mirrors the Python
    # client's deliberate scope boundary (see PYTHON_R_INTEROP_PLAN.md
    # checkpoint 1 findings). Fail loudly rather than guess the shape.
    stop(linal_error(paste(
      "TensorTable unwrapping is not yet implemented (no verified wire",
      "example) -- see PYTHON_R_INTEROP_PLAN.md checkpoint 3 findings"
    )))
  }
  if (kind %in% c("Tensor", "LazyTensor")) {
    return(structure(
      list(
        shape = as.integer(unlist(payload$shape$dims)),
        data = as.numeric(unlist(payload$data))
      ),
      class = "linal_tensor_result"
    ))
  }
  stop(linal_error(paste0("Unknown DslOutput variant: ", kind)))
}

File: clients/python/linaldb_server/wire.py
"""Unwraps the JSON shapes documented in ``clients/CONTRACT.md`` into
plain Python values. Kept separate from ``client.py`` so it can be unit
tested against fixture JSON with no HTTP involved.
"""

from .errors import LinalError

_SCALAR_UNWRAPPERS = {
    "Float": float,
    "Int": int,
    "String": str,
    "Bool": bool,
}


def unwrap_value(value):
    """Unwrap one tagged ``Value`` cell (contract §3) into a plain Python
    value: float/int/str/bool/None, ``list[float]`` for a Vector, or
    ``list[list[float]]`` for a Matrix.
    """
    if value == "Null":
        return None
    if isinstance(value, dict) and len(value) == 1:
        (key, inner), = value.items()
        if key in _SCALAR_UNWRAPPERS:
            return _SCALAR_UNWRAPPERS[key](inner)
        if key == "Vector":
            return [float(x) for x in inner]
        if key == "Matrix":
            return [[float(x) for x in row] for row in inner]
    raise LinalError(f"Unrecognized Value wire shape: {value!r}")


class TableResult:
    """A `Table` result: `.columns` (list of names, in order) and
    `.rows` (list of tuples of already-unwrapped Python values, one
    tuple per row, in column order).
    """

    def __init__(self, columns, rows):
        self.columns = columns
        self.rows = rows

    def __repr__(self):
        return f"TableResult(columns={self.columns!r}, rows={len(self.rows)})"

    @classmethod
    def from_wire(cls, payload):
        # See CONTRACT.md's verified `Table` shape: top-level `schema`
        # gives column order/names once; each row's cells live under a
        # `values` key (not the row object itself).
        columns = [f["name"] for f in payload["schema"]["fields"]]
        rows = [
            tuple(unwrap_value(v) for v in row["values"])
            for row in payload["rows"]
        ]
        return cls(columns, rows)


def _default_strides(shape):
    """Row-major (C-order) contiguous strides for `shape`, in element
    counts -- matching the engine's `Tensor::compute_default_strides`
    (src/core/tensor.rs): the last dimension has stride 1, accumulating
    right-to-left. Used when the wire payload omits `strides` (the
    assume-contiguous case).
    """
    strides = [0] * len(shape)
    stride = 1
    for i in reversed(range(len(shape))):
        strides[i] = stride
        stride *= shape[i]
    return strides


class TensorResult:
    """A standalone `Tensor`/`LazyTensor` result. Structural shape
    verified against a live server for `shape`/`strides`/`offset`/`data`
    (see `test_transpose_over_http_to_numpy_end_to_end` in
    `test_client_integration.py`) -- CONTRACT.md's original caveat about
    this not being independently verified applied to `TableResult` only
    by the time this was checked.

    `.to_numpy()` honors `strides`/`offset` exactly as the engine defines
    them (element counts, row-major -- see `src/core/tensor.rs`), so a
    zero-copy result like `TRANSPOSE` (which swaps strides and reuses its
    input's buffer as-is, never copying data) reconstructs correctly.
    """

    def __init__(self, shape, data, strides=None, offset=0):
        self.shape = shape
        self.data = data
        self.strides = strides
        self.offset = offset

    def __repr__(self):
        return f"TensorResult(shape={self.shape!r}, len(data)={len(self.data)})"

    @classmethod
    def from_wire(cls, payload):
        return cls(
            shape=payload["shape"]["dims"],
            data=payload["data"],
            strides=payload.get("strides"),
            offset=payload.get("offset", 0),
        )

    def to_numpy(self):
        """Reconstruct this tensor's logical values as an owned
        `numpy.ndarray`. Needed because a zero-copy engine op like
        `TRANSPOSE` (src/engine/kernels.rs) swaps `strides` and reuses
        the pre-transpose buffer as-is -- a plain `reshape` of `data`
        would silently reinterpret the raw buffer in the wrong order
        instead of raising, since it never looks at `strides`/`offset`
        at all.
        """
        import numpy as np

        flat = np.asarray(self.data, dtype="float32")
        shape = tuple(self.shape)
        strides = self.strides if self.strides is not None else _default_strides(shape)

        if len(strides) != len(shape):
            raise LinalError(
                f"Tensor wire payload shape/strides rank mismatch: "
                f"shape={shape!r} strides={strides!r}"
            )

        offset = self.offset or 0
        # Fail loudly instead of letting as_strided silently read past
        # the buffer on a malformed/mismatched payload.
        max_index = offset
        for dim, stride in zip(shape, strides):
            if dim > 0:
                max_index += (dim - 1) * stride
        if flat.size and max_index >= flat.size:
            raise LinalError(
                f"Tensor wire payload out of bounds: shape={shape!r} "
                f"strides={strides!r} offset={offset!r} but data has "
                f"only {flat.size} element(s)"
            )

        itemsize = flat.itemsize
        view = np.lib.stride_tricks.as_strided(
            flat[offset:],
            shape=shape,
            strides=tuple(s * itemsize for s in strides),
            writeable=False,
        )
        # Detach from the raw/possibly-oversized/shared buffer -- callers
        # get a normal, fully-owned array.
        return view.copy()


def unwrap_result(result):
    """Unwrap one `DslOutput` JSON value (contract §1) into `None` (no
    output), a plain `str` (`Message`), a `TableResult`, or a
    `TensorResult`.
    """
    if result is None:
        return None
    if not isinstance(result, dict) or len(result) != 1:
        raise LinalError(f"Unrecognized DslOutput wire shape: {result!r}")
    (kind, payload), = result.items()
    if kind == "Message":
        return payload
    if kind == "Table":
        return TableResult.from_wire(payload)
    if kind == "TensorTable":
        # Not yet exercised against a real response — no confirmed
        # example command produces one in this checkpoint's test pass.
        # Fail loudly rather than guess at the wire shape.
        raise LinalError(
            "TensorTable unwrapping is not yet implemented (no verified "
            "wire example) — see PYTHON_R_INTEROP_PLAN.md checkpoint 1 findings"
        )
    if kind in ("Tensor", "LazyTensor"):
        return TensorResult.from_wire(payload)
    raise LinalError(f"Unknown DslOutput variant: {kind!r}")
